### PR TITLE
Aura/drtvwr 558 dbljmp

### PR DESCRIPTION
**Pull request by auralinden**
Updated the AgentIO module and doubleleap.py example to reflect syntax changes.

### CODE REVIEW

**Review by nat-goodspeed — COMMENTED** [review]
I see that the new API is all asynchronous. That is, when a client calls `request_lookat()`, there's no expectation that it can immediately call `getLookAt()`: presumably some amount of real time will elapse before the viewer sends a `"look_at"` response whose handler will set `_look_at`. I assume some (most?) clients will end up coding a spin loop, repeatedly calling `getLookAt()` until it returns something that seems reasonable.

This is why I'm a fan of using eventlet coroutines and a stdin handler that matches `reqid` values. In VITA and its derivatives, you can code a client request for some data, block waiting for that response and return the response data to your caller. But only the calling coroutine is blocked; the other client coroutines continue running.

I have a pending TODO to merge the VITA and viewer updater machinery into the leap repo, so that both VITA and the viewer updater will rely on the leap repo as the One Source of Truth for Python LLLeap client scripts. Maybe I should accelerate that.

One other comment about `sendLeapRequest()`: it seems as though the function signature could/should be

    def sendLeapRequest(self, command, data={}):

so that instead of:

    msg = { 'command':'get_lookat', 'data':{} }
    _controller.sendLeapRequest(msg)

you could simply write:

    _controller.sendLeapRequest('get_lookat')

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py]: If this script isn't going to be executed directly, it doesn't need a #!.

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py]: These license headers have been removed with LEAP moving to MIT, per [updated coding policy](https://wiki.secondlife.com/wiki/Coding_standard#MIT_Licensed_Code).

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:14]: Are these globals necessary, or could they be kept in a class instance? I'd suggest refactoring this file to keep these values in a class, which can be initialized where it is consumed.

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:16]: nit: PEP-8: naming

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:28]: nit: PEP-8: naming

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:19]: Nit: this doesn't need to start with a `_` since it's not private --it's enclosed in the function scope. You might also want to use the `functools.wraps` helper to preserve parent function metadata.

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:39]: nit: PEP-8: naming. Last one I'll mention, but there are others.

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:55]: Is there a more specific error this should be catching?

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py:58]: Here too, should this be excepting a more specific error?

File: agentio/__init__.py
Comment (bennettgoble) [inline agentio/__init__.py]: I don't think this warning/exception is necessary.

File: puppetry/__init__.py
Comment (bennettgoble) [inline puppetry/__init__.py]: Coding standard: don't need this header.

File: puppetry/__init__.py
Comment (bennettgoble) [inline puppetry/__init__.py:1]: Don't need this if it's not going to be run.

File: puppetry/__init__.py
Comment (bennettgoble) [inline puppetry/__init__.py]: nit: Trailing whitespace. Should some linter be added for this or does pre-commit already cover it?

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py]: This pass isn't necessary.

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py]: Is there a more specific exception this could be catching?

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py:151]: Pass `type=int` to `add_argument()` to have the type initialized.

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py:129]: What exceptions are caught/ignored here?

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py:132]: nit: You're using both `'` and `"` in docstrings in this file.

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py]: Returning here skips `puppetry.stop()` --should that be getting called?

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py:120]: `puppetry.stop()` is also called after the loop ends --is that duplication harmful?

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py]: Is glm used? Be sure to clean up any unused imports.

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py:19]: Couple loggers getting thrown around: module-scoped logger and puppetry's logger. Do we need both?

File: puppetry/examples/doubleleap.py
Comment (bennettgoble) [inline puppetry/examples/doubleleap.py:45]: Nit: strings in this file are inconstantly using both `'` and `"`.

**Review by bennettgoble — CHANGES_REQUESTED** [review]
Nice work with the new demo, @auralinden. I have some suggested changes and a couple questions:

- This PR adds a handful of empty files. Are these needed?
- The style of the agentio module is a bit odd for python: I think this could be refactored to be a bit more object oriented, which would reduce use of global state/singletons and make any future testing easier and module-reuse possible
- I've added a handful of comments about python coding style: would it be helpful to add some linters flake8 and/or pylint to catch issues as part of commit hooks?
- Looks like the PR is failing its commit hooks (pre-commit) could you install pre-commit and run these hooks, address the issues?

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:14]: I don't like the look either but it is consistent with the leap/__init__.py and puppetry/__init__.py scripts
I did some quick tests and found that the globals are file-scoped and are effectively acting as class variables.   They are not accessible by a bare global _controller in another file but would instead referenced as agentio._controller

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:16]: same pattern as puppetry and leap implementations.
I'd prefer to change all or none of them.
Changing the other modules and demo script is easy but I'll run it by the puppetry group.

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:55]: Fixed.

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:58]: Fixed

File: puppetry/examples/doubleleap.py
Comment (auralinden) [reply puppetry/examples/doubleleap.py:120]: It's fine.

File: agentio/LICENSE.md
Comment (bennettgoble) [inline agentio/LICENSE.md]: There's already a LICENSE in the root of the project. Don't think another one is needed.

File: agentio/__init__.py
Comment (bennettgoble) [reply agentio/__init__.py:55]: Hmm, I don't think this addresses the suggestion. `except Exception as e` will catch all exceptions, which is the same behavior as before. Is there a more specific class or classes of expected exceptions?

**Review by bennettgoble — CHANGES_REQUESTED** [review]

File: puppetry/examples/doubleleap.py
Comment (auralinden) [reply puppetry/examples/doubleleap.py:129]: Unclear.  This pattern came out of puppetry.  I don't see any examples of it being used this way in a handful of examples I looked at and found nothing useful searching for exceptions to this function.  I'm going to leave it for now.

File: puppetry/examples/doubleleap.py
Comment (auralinden) [reply puppetry/examples/doubleleap.py]: I don't really see a way this can fail so I've removed the exception handler.

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:16]: Going to do these as a separate PR (pending further discussion)

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:19]: This and other syntax changes will be handled in a separate PR.

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:28]: This and other syntax changes will be handled in a separate PR.

File: agentio/__init__.py
Comment (auralinden) [reply agentio/__init__.py:39]: This and other syntax changes will be handled in a separate PR.

**Review by auralinden — COMMENTED** [review]
Resolved issues except for the items that are marked as being handled in a separate PR.

File: agentio/LICENSE.md
Comment (auralinden) [reply agentio/LICENSE.md]: Deleted

**Review by bennettgoble — APPROVED** [review]
Spoke with @auralinden in slack about keeping API changes separate from this work adding demos. With that sorted, LGTM.

Looks like [pre-commit](https://pre-commit.com/) is still failing, so you'll need to run its hooks and commit any changes before merging.

**Review by auralinden — COMMENTED** [review]
Fixed pre-commit.  Can I approve now?  Looks like no.  Maybe adding a comment will help.

**Pull request merged**
